TL/MLX5: custom ib wqes

samnordmann opened the pull request:
## What
Implements the WQEs for transpose, wait_on_data, umr, and rdma

vspetrov commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,301 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2022.  ALL RIGHTS RESERVED.
```
2023

vspetrov commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,301 @@
+    eb               = rb->entries;
+
+    /*
+	 * ------------------------------------------------------------
```
comment alignment

samnordmann replied:
on my editor the stars are aligned. I think that's just maybe a space character that is rendered as a tab in github. I could align the stars with the slash "/"

samnordmann replied:
solved

vspetrov commented on src/components/tl/mlx5/tl_mlx5_wqe.h:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2022.  ALL RIGHTS RESERVED.
```
2023

vspetrov reviewed (commented):
We should implement gtests for these functions (there are already some ib gtests available).

For transpose:
init 1 qp, connect loop back.
allocate some send and recv buffers (matrices). for example, ncols=32, nrows=32, element = int. Initialize the matrices. Do buffers registratoin. Post transpose WQE. poll completion. check matrix is tranpsposed correctly.

Umr:
simple test: just allocate "strided" pattern with UMR, perform single ibv_post_send with rdma write and check the pattern correctly moved.

Post rdma: simple comparison of ucc_tl_mlx5_post_rdma with ibv_post_send with rdma opcode

Wait on data:
allocate buffer of 3 INTs (int buffer[3]). Initialize it to 0. Register.
Same single local QP connected to self (loopback).
Post wait on data addr = &buffer[0]; value = 1;
Post rdma write src = &buffer[1]; dst = &buffer[2];

do local assignment from host buffer[1] = 0xdeadbeef;
(memory barrier)
do local assignment from host buffer[0] = 1;
poll completion of 2nd rdma_write.
check that buffer[2] gets 0xdeadbeef;

Later we can add more stresfull tests (more complex UMR patterns, including nested mkeys; more complex transpose patterns - different ncols, nrows, elem types)

samnordmann commented:
> We should implement gtests for these functions (there are already some ib gtests available).
> For transpose: init 1 qp, connect loop back. allocate some send and recv buffers (matrices). for example, ncols=32, nrows=32, element = int. Initialize the matrices. Do buffers registratoin. Post transpose WQE. poll completion. check matrix is tranpsposed correctly.
 
Thank you very much @vspetrov for those instructions. I'm working on the Gtests. Here is the first version of the test for the transpose WQE.
I will now implement the other tests you describe

samnordmann commented:
@vspetrov The tests have been added as you suggested. Few comments:
- Wait-on-Data: there was a small mistake on the `data_mask` field of the WQE. This field should be set to 1 but was previously set to 0. This has been fixed.
- UMR: the test passes for a very wide range of parameters, except when the number of sources is a multiple of 4. I need to figure out what happens in this particular case. For now, those cases are simply skipped.

Sergei-Lebedev commented on test/gtest/tl/mlx5/test_tl_mlx5_wqe.cc:
```diff
@@ -0,0 +1,317 @@
+    umr_mkey_init_attr.create_flags = MLX5DV_MKEY_INIT_ATTR_FLAGS_INDIRECT;
+    umr_mkey_init_attr.max_entries  = nbr_srcs;
+
+    umr_mkey = mlx5dv_create_mkey(&umr_mkey_init_attr);
```
check max_entries is greater or equal than nbr_srcs on function return

samnordmann replied:
I am not sure to understand.
I added a check that the qp's `max_inline_data` is greater than `nbr_srcs`. Does it address your comment ?

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,304 @@
+    uint32_t                  opcode = MLX5_OPCODE_WAIT;
+    uint32_t                  opmode = 0x1; //wait on data
+    uint32_t                  n_ds   = 3;   //CTRL + Wait on Data of Size 2
+    char                      wqe_desc[n_ds * DS_SIZE];
```
initialized vars go first

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,304 @@
+                                  uint32_t ptr_mkey, void *ptr_address)
+{
+    uint32_t opcode = MLX5_OPCODE_UMR;
+    uint64_t reglen = 0;
```
fix alignment

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,304 @@
+ucc_status_t ucc_tl_mlx5_post_umr(struct ibv_qp *     qp,
+                                  struct mlx5dv_mkey *dv_mkey,
+                                  uint32_t access_flags, uint32_t repeat_count,
+                                  uint16_t                      num_entries,
```
remove extra spaces

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,304 @@
+ * | repeat_block | entry_block | entry_block |...| entry_block |
+ * While the repeat entry contains details on the list of the block_entries. */
+static void umr_pointer_seg_init(uint32_t                      repeat_count,
+                                 uint16_t                      num_interleaved,
```
extra spaces

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,304 @@
+
+#define ACTION_RETRY        0x0ULL
+#define ACTION_SEND_ERR_CQE 0x1ULL
+#define ACTION           ACTION_RETRY
```
alignment

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.h:
```diff
@@ -0,0 +1,37 @@
+
+#ifndef UCC_TL_MLX5_WQE_H_
+#define UCC_TL_MLX5_WQE_H_
+#include "ucc/api/ucc_status.h"
```
blank line before include

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,304 @@
+    struct mlx5_wqe_data_seg *    data;
+    struct mlx5_wqe_datagram_seg *dseg;
+    struct mlx5_wqe_raddr_seg *   rseg;
+    uint32_t n_ds = (sizeof(*ctrl) + (ah ? sizeof(*dseg) : 0) + sizeof(*rseg) +
```
initialized vars first

Sergei-Lebedev replied:
just a code style we stick to, if you need some uninit variables then please move n_ds initialization out of declaration block

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,304 @@
+    uint32_t                      opmode = 0x0;
+    struct ibv_qp_ex *            qp_ex = ibv_qp_to_qp_ex(qp);
+    struct mlx5dv_qp_ex *         mqp   = mlx5dv_qp_ex_from_ibv_qp_ex(qp_ex);
+    uint8_t                       fm_ce_se = MLX5_WQE_CTRL_INITIATOR_SMALL_FENCE;
```
alignment

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,304 @@
+        tl_mlx5_ah_to_av(ah, &dseg->av);
+        dseg->av.dqp_dct |= htobe32(qpn | MLX5_EXTENDED_UD_AV);
+        dseg->av.key.dc_key = htobe64(DC_KEY);
+
```
rem blank line

Sergei-Lebedev commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,304 @@
+    char                      wqe_desc[n_ds * DS_SIZE];
+    struct ibv_qp_ex *        qp_ex = ibv_qp_to_qp_ex(qp);
+    struct mlx5dv_qp_ex *     mqp   = mlx5dv_qp_ex_from_ibv_qp_ex(qp_ex);
+    uint8_t fm_ce_se = MLX5_WQE_CTRL_FENCE | MLX5_WQE_CTRL_CQ_UPDATE;
```
alignment

samnordmann commented on test/gtest/tl/mlx5/test_tl_mlx5_qps.cc:
```diff
@@ -0,0 +1,47 @@
+ * See file LICENSE for terms.
+ */
+
+#include "test_tl_mlx5_qps.h"
```
Currently I think that the tests are not run. I doubt even than the tl/mlx5 is built
Let me know what to do about it. This requires the node to have CX7

Sergei-Lebedev commented on test/gtest/Makefile.am:
```diff
@@ -106,7 +106,9 @@ gtest_SOURCES =                     \
 if TL_MLX5_ENABLED
-    gtest_SOURCES  += tl/mlx5/test_tl_mlx5.cc
+    gtest_SOURCES  += tl/mlx5/test_tl_mlx5.cc \
+					  tl/mlx5/test_tl_mlx5_qps.cc\
```
it's not GH, you have mix of spaces and tabs in that snippet

![image](https://user-images.githubusercontent.com/5495223/223936280-e94ca6a8-adce-4faf-9814-a5ad2dece1a9.png)

bureddy commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,306 @@
+    char                              wqe_desc[n_ds * DS_SIZE];
+    int                               xlat_size;
+
+    ibv_wr_start(qp_ex);
```
do you need to protect all the below code here? or you can move ibv_wr_start() just before WQE builder mlx5dv_wr_raw_wqe at line 195 to reduce the critical section?

bureddy commented on src/components/tl/mlx5/tl_mlx5_wqe.c:
```diff
@@ -0,0 +1,306 @@
+    if (ah) {
+        dseg = PTR_OFFSET(ctrl, sizeof(*ctrl));
+        tl_mlx5_ah_to_av(ah, &dseg->av);
+        dseg->av.dqp_dct |= htobe32(qpn | MLX5_EXTENDED_UD_AV);
```
|= alignment

bureddy commented on test/gtest/tl/mlx5/test_tl_mlx5_wqe.cc:
```diff
@@ -0,0 +1,312 @@
+    GTEST_ASSERT_EQ(wcs[0].wr_id, 0);
+
+    // RDMA Write
+    ibv_wr_start(qp.qp_ex);
```
Why not ibv_wr_start and ibv_wr_complete() part of ucc_tl_mlx5_post_rdma()

vspetrov replied:
> Why not ibv_wr_start and ibv_wr_complete() part of ucc_tl_mlx5_post_rdma()

i think because we can post multiple outstanding rmda to the same peer in a batch and make a single critical section for them (it will be in the send_block fn of the algorithm)

bureddy commented on test/gtest/tl/mlx5/test_tl_mlx5_wqe.cc:
```diff
@@ -0,0 +1,312 @@
+        ibv_reg_mr(pd, dst, nrows * ncols * elem_size, IBV_ACCESS_LOCAL_WRITE);
+    GTEST_ASSERT_NE(nullptr, dst_mr);
+
+    ibv_wr_start(qp.qp_ex);
```
Why not ibv_wr_start and ibv_wr_complete() part of ucc_tl_mlx5_post_transpose()

bureddy commented on test/gtest/tl/mlx5/test_tl_mlx5_qps.h:
```diff
@@ -0,0 +1,175 @@
+                                                   struct ibv_ah **ah_ptr,
+                                                   ucc_base_lib_t *lib);
+
+class test_tl_mlx5_dc : public test_tl_mlx5_qp {
```
is "test_tl_mlx5_dc" class used some where? i dont see it

samnordmann replied:
This class is used for the tests in `test_tl_mlx5_qps.cc`
For now the tests on dc are only about qp creation. We will add more tests later when we will implement support for dc qps (for now, only rc qps are fully supported)

samnordmann commented:
@bureddy I moved all the calls to `ibv_wr_start` inside the WQE post functions, as you suggest

vspetrov commented:
> @bureddy I moved all the calls to `ibv_wr_start` inside the WQE post functions, as you suggest

then you'll need to carefully fix the upcoming "algorithm PR". There the logic is: 
wr_start
do several post_transpose + post_rdma
wr_end
now, you will have recursive "ibv_wr_start"

bureddy commented:
> > @bureddy I moved all the calls to `ibv_wr_start` inside the WQE post functions, as you suggest
> 
> then you'll need to carefully fix the upcoming "algorithm PR". There the logic is: wr_start do several post_transpose + post_rdma wr_end now, you will have recursive "ibv_wr_start"

samnordmann commented:
> > @bureddy I moved all the calls to `ibv_wr_start` inside the WQE post functions, as you suggest
> 
> then you'll need to carefully fix the upcoming "algorithm PR". There the logic is: wr_start do several post_transpose + post_rdma wr_end now, you will have recursive "ibv_wr_start"

Right @vspetrov, according to what you are saying, I reverted the last change and kept the ibv_wr_start/complete outside RDMA write and Transpose WQE.
@bureddy is it ok for you ?

bureddy commented:
> > > @bureddy I moved all the calls to `ibv_wr_start` inside the WQE post functions, as you suggest
> > 
> > 
> > then you'll need to carefully fix the upcoming "algorithm PR". There the logic is: wr_start do several post_transpose + post_rdma wr_end now, you will have recursive "ibv_wr_start"

ok. will recheck with alg PR. we have to keep the critical section with wr_start/wr_complete should be a minimum..

samnordmann commented:
bot:retest

samnordmann commented:
bot:retest

samnordmann commented:
bot:retest
